### PR TITLE
Add dialog to enter credentials on errored mount point

### PR DESCRIPTION
**Pull request by jmaciasportela**
To be merged after https://github.com/owncloud/core/pull/20569 because is a continuation

### CODE REVIEW

**Comment by Xenopathic** [comment]
@jmaciasportela Please don't merge from master into a PR, rather rebase on top of master. So instead of running `git merge master`, run `git rebase master` (and hope you get no conflicts...)

**Comment by jmaciasportela** [comment]
@Xenopathic I didn't do that. I create branch issue_20888 from branch issue_20427 because I needed to continue with that code and is not merged on master yet. Morris explain me how to do rebase so I will do in a future. Thanks.

**Comment by icewind1991** [comment]
Rebased on #21790 using the proper endpoints

**Comment by PVince81** [comment]
It works!!

**Comment by PVince81** [comment]
:+1: from me

**Comment by SergioBertolinSG** [comment]
Red row has a visual bug, perhaps we can open it after merging this:

Click on the center of the row and dragging, turns a part of the row white.

screen shot 2016-01-27 at 17 40 48

**Comment by SergioBertolinSG** [comment]
According to https://github.com/owncloud/core/issues/21911#issuecomment-175104752

It should be named "User entered, store in database" in the dropdown.

**Comment by SergioBertolinSG** [comment]
Enter key doesn't work for saving credentials. You have to click the save button with the mouse.

screen shot 2016-01-27 at 17 52 01

**Comment by SergioBertolinSG** [comment]
There is no way to change credentials if they work. This is not expected to happen in any situation?

**Comment by icewind1991** [comment]
> There is no way to change credentials if they work. This is not expected to happen in any situation?

You can change them in personal settings

**Comment by SergioBertolinSG** [comment]
> You can change them in personal settings

The mount is system wide, created by admin. User entered correct credentials (row is white now). But personal settings nothing appears to change it.

**Comment by PVince81** [comment]
@icewind1991 please rebase to include the merged changes from the other branch.
There are conflicts.

File: apps/files_external/js/statusmanager.js [hunk 1]
```diff
@@ -163,21 +168,25 @@ OCA.External.StatusManager = {
...
 				var mountData = allMountStatus[name];
 				if (mountData.type === "system") {
-					OC.dialogs.confirm(t('files_external', 'There was an error with message: ') + mountData.error + '. Do you want to review mount point config in admin settings page?', t('files_external', 'External mount error'), function(e){
-						if(e === true) {
-							window.location.href = OC.generateUrl('/settings/admin#files_external');
-						}
-					});
+					if (mountData.userProvided) {
+						// personal mount whit credentials problems
+						this.showCredentialsDialog(name, mountData, 'fakesavecredentials.php');
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: @icewind1991 @jmaciasportela "fakesavecredentials.php" ?
Needs to be replaced with the new endpoint (or removed)

File: apps/files_external/js/statusmanager.js [hunk 2]
```diff
@@ -163,21 +168,25 @@ OCA.External.StatusManager = {
...
+					} else {
+						OC.dialogs.confirm(t('files_external', 'There was an error with message: ') + mountData.error + '. Do you want to review mount point config in admin settings page?', t('files_external', 'External mount error'), function (e) {
+							if (e === true) {
+								window.location.href = OC.generateUrl('/settings/admin#files_external');
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: Please use `OC.redirect()` instead, it makes it unit-testable.

File: apps/files_external/js/statusmanager.js [hunk 3]
```diff
@@ -163,21 +168,25 @@ OCA.External.StatusManager = {
...
+							}
+						});
+					}
 				} else {
-					OC.dialogs.confirm(t('files_external', 'There was an error with message: ') + mountData.error + '. Do you want to review mount point config in personal settings page?', t('files_external', 'External mount error'), function(e){
-						if(e === true) {
-							window.location.href = OC.generateUrl('/settings/personal#external-storage');
+					OC.dialogs.confirm(t('files_external', 'There was an error with message: ') + mountData.error + '. Do you want to review mount point config in personal settings page?', t('files_external', 'External mount error'), function (e) {
+						if (e === true) {
+							window.location.href = OC.generateUrl('/settings/personal#' + t('files_external', 'goto-external-storage'));
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: `OC.redirect()`

Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: Please remove the "goto-", see https://github.com/owncloud/core/pull/21438

File: apps/files_external/js/statusmanager.js [hunk 4]
```diff
@@ -355,12 +366,84 @@ OCA.External.StatusManager = {
 
 		self.processMountList(mountListData);
 		self.launchPartialConnectivityCheck(mountListData, recheck);
+	},
+
+	/**
+	 * Function to display custom dialog to enter credentials
+	 * @param mountPoint
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: `{Object}` mountPoint

When using JSDoc structures, ideal is to make classes.
Else you can also write JSDoc-only objects to describe it with `typedef`, see https://github.com/owncloud/core/blob/32067ac49b6f891471a6b5672c903a390b4ac26d/apps/files/js/fileactions.js#L677

File: apps/files_external/js/statusmanager.js [hunk 5]
```diff
@@ -355,12 +366,84 @@ OCA.External.StatusManager = {
...
+					$popup.ocdialog('close');
+					/* Trigger status check again */
+					OCA.External.StatusManager.recheckConnectivityForMount([OC.basename(data.mountPoint)], true);
+				},
+				error: function () {
+					$('.oc-dialog-close').show();
+					OC.Notification.showTemporary(t('files_external', 'Credentials saving failed'));
+				}
+			});
+			console.log('apply', username, password, endpoint);
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: Remove debug statement

File: apps/files_external/js/statusmanager.js [hunk 6]
```diff
@@ -355,12 +366,84 @@ OCA.External.StatusManager = {
...
 		self.launchPartialConnectivityCheck(mountListData, recheck);
+	},
+
+	/**
+	 * Function to display custom dialog to enter credentials
+	 * @param mountPoint
+	 * @param mountData
+	 */
+	showCredentialsDialog: function (mountPoint, mountData) {
+		var $popup = $('<div id="files_external_div_form"/>').attr('title', t('files_external', 'Mount Credentials'));
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: I know this is probably old code. In the future, please consider using Handlebar templates, see https://github.com/owncloud/core/blob/32067ac49b6f891471a6b5672c903a390b4ac26d/core/js/sharedialogview.js#L16 and https://github.com/owncloud/core/blob/32067ac49b6f891471a6b5672c903a390b4ac26d/core/js/sharedialogview.js#L347

File: apps/files_external/js/statusmanager.js [hunk 7]
```diff
@@ -407,7 +490,7 @@ OCA.External.StatusManager.Utils = {
 		}
 		trFolder.removeClass('externalErroredRow').removeClass('externalDisabledRow');
 		tdChilds = trFolder.find("td:first-child div.thumbnail");
-		tdChilds.each(function(){
+		tdChilds.each(function () {
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: Please don't reformat everything automatically in the future

Comment (jmaciasportela) [inline apps/files_external/js/statusmanager.js]: ACK

File: apps/files_external/js/statusmanager.js (hunk 6 above)
Comment (jmaciasportela) [inline apps/files_external/js/statusmanager.js]: ACK

**Comment by icewind1991** [comment]
Switched the dialog to handlebars

**Comment by PVince81** [comment]
Wow, nice teamwork :wink:

**Comment by PVince81** [comment]
Let me know when you are done and set the status to "To review". Thanks.

**Comment by icewind1991** [comment]
good for review imo

**Comment by PVince81** [comment]
Steps:

1. Create a SFTP storage, enter credentials, etc
1. Now switch it to "User Provided", wait for the yellow status
1. Go to "All files"
1. => ownCloud is unusable, getting this:

no-credentials-saved

When I switch to master, I can access the file list.
So this branch seems to introduce this issue.

**Comment by jmaciasportela** [comment]
Tested with other storage backend, same result.

Nothing in logs...

Response to ajax call when change auth method:
```json
{

    "id":1,
    "mountPoint":"\/SMB",
    "backend":"smb",
    "authMechanism":"password::userprovided",
    "backendOptions":{
        "host":"10.10.10.106",
        "share":"shared",
        "root":"\/",
        "domain":"",
        "user":"administrador",
        "password":"SolidGear-SL1"
    },
    "priority":100,
    "mountOptions":{
        "encrypt":true,
        "previews":true,
        "filesystem_check_changes":1
    },
    "status":2,
    "statusMessage":"Insufficient data: No credentials saved",
    "userProvided":true,
    "type":"system"

}
```

File: apps/files_external/lib/config/configadapter.php [hunk 8]
```diff
@@ -130,6 +130,11 @@ public function getMountsForUser(IUser $user, IStorageFactory $loader) {
 				$impl = new FailedStorage(['exception' => $e]);
 			}
 
+			$availability = $impl->getAvailability();
+			if (!$availability['available']) {
+				$impl = new FailedStorage(['exception' => null]);
```
Comment (PVince81) [inline apps/files_external/lib/config/configadapter.php]: Maybe this is the part that introduces that global exception ?

Comment (icewind1991) [inline apps/files_external/lib/config/configadapter.php]: That path is triggered if the storage has been configured correctly, then got it's credentails changes to be invalid (or probably also if the backend server dies or something)

**Comment by PVince81** [comment]
@jmaciasportela any update on why this breaks everything ?

**Comment by jmaciasportela** [comment]
No. No logs, so I'm trying to have xdebug working on docker container in docker.oc.solidgear.es but is not working yet.

File: apps/files_external/lib/config/configadapter.php (hunk 8 above)
Comment (PVince81) [inline apps/files_external/lib/config/configadapter.php]: If I remove this hunk the error disappears

File: apps/files_external/js/statusmanager.js [hunk 9]
```diff
@@ -355,12 +366,85 @@ OCA.External.StatusManager = {
...
 		self.processMountList(mountListData);
 		self.launchPartialConnectivityCheck(mountListData, recheck);
+	},
+
+	credentialsDialogTemplate:
+		'<div id="files_external_div_form"><div>' +
+		'<div>{{credentials_text}}</div>' +
+		'<form>' +
+		'<input type="text" name="username" placeholder="{{placeholder_username}}"/>' +
+		'<input type="text" name="password" placeholder="{{placeholder_password}}"/>' +
```
Comment (PVince81) [inline apps/files_external/js/statusmanager.js]: OUCH, this must be `type="password"`

File: apps/files_external/lib/config/configadapter.php (hunk 8 above)
Comment (PVince81) [inline apps/files_external/lib/config/configadapter.php]: and everything else works fine.

@icewind1991 can you help fixing the code path you described where the credentials become invalid ?

File: apps/files_external/js/statusmanager.js (hunk 9 above)
Comment (icewind1991) [inline apps/files_external/js/statusmanager.js]: fixed

**Comment by PVince81** [comment]
also the dialog's title is "undefined".
Would also be good if the enter key would trigger Save, but not sure if it's possible with ocdialogs.

**Comment by PVince81** [comment]
Now fails with:
```
{"reqId":"3rAJTomxHKn0NNM2LJcu","remoteAddr":"127.0.0.1","app":"PHP","message":"Class OCA\\Files_External\\Lib\\FailedCache contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (OCP\\Files\\Cache\\ICache::insert) at \/srv\/www\/htdocs\/owncloud\/apps\/files_external\/lib\/failedcache.php#126","level":3,"time":"2016-02-04T15:30:49+00:00","method":"GET","url":"\/owncloud\/index.php"}
```

**Comment by icewind1991** [comment]
should be fixed with rebase

**Comment by PVince81** [comment]
ownCloud doesn't brick itself anymore, but now the status indicator is always red for SFTP, not yellow.
And then in the files list there is no row for the storage, so I can't click it.

Let me add my exact steps.

**Comment by PVince81** [comment]
1. Reset ownCloud / run setup
1. Enable ext storage app
1. Go to admin page
1. Type "sftp" as mount name
1. Select "SFTP" as mount tye
1. Select "User entered, ..." auth backend
1. Type host name
1. Tab, Type root path (do this before it gets a chance to save, at this stage there is no status indicator)
1. Click outside the fields, wait for save

Expected: yellow indicator
Actual: red indicator

No exceptions or stack traces unfortunately.

@icewind1991 what are your exact steps ?

**Comment by PVince81** [comment]
same issue with SMB, not specific to SFTP

**Comment by icewind1991** [comment]
@PVince81 fixed

**Comment by PVince81** [comment]
The storage logic works now.

Let me try and fix the dialog

**Comment by PVince81** [comment]
Here we go, the dialog looks better now.

@icewind1991 @jmaciasportela please review my commit.

:+1: for all changes from @jmaciasportela and @icewind1991

**Comment by icewind1991** [comment]
:+1:

**Comment by PVince81** [comment]
Needs a second reviewer for the latest changes @Xenopathic @jmaciasportela @MorrisJobke @rullzer

**Comment by PVince81** [comment]
steps are here https://github.com/owncloud/core/pull/20920#issuecomment-179904766

**Comment by jmaciasportela** [comment]
Works Fine ! :+1:

**Comment by SergioBertolinSG** [comment]
https://github.com/owncloud/core/pull/20920#issuecomment-175736933 this has been opened  separately here https://github.com/owncloud/enterprise/issues/1146

**Pull request merged**